src/hierarchical.rs

```diff
@@ -0,0 +1,244 @@
+pub struct Agglomerative<'a, P: 'a + Point> {
+    points: Vec<&'a P>,
+    distances: Vec<Vec<f64>>
+}
+
+impl<'a, P: 'a + Point> Agglomerative<'a, P> {
+    /// Pre-compute the inter-point distances of `data`.
+    pub fn new<T>(data: T) -> Agglomerative<'a, P>
+        where T: IntoIterator<Item=&'a P>
+    {
+        let points:Vec<_> = data.into_iter().collect();
+        let n = points.len();
+
+        // pre-compute distances, using as few space as possible,
+        // we assume that d(i,j) = d(j,i) and d(i,i) = 0
+        let distances = (0..n).map(|i| {
+            (i+1..n).map(|j| {
+                points[i].dist(points[j])
+            }).collect()
+        }).collect();
+
+        Agglomerative {
+            points: points,
+            distances: distances
+        }
+    }
+
+    /// Perform the actual clustering.
+    ///
+    /// Starting with each point in its own cluster successively merge two clusters
+    /// ,determined by `linkage`, until either only one is left or the maximal distance
+    /// `threshold` has been exceeded.
+    pub fn clusters(&'a self, linkage: LinkageCriterion, threshold: f64) -> Vec<Vec<&'a P>> {
+        self.merge(linkage, threshold).iter().map(|cluster| {
+            cluster.into_iter().map(|i| self.points[*i]).collect()
+        }).collect()
+    }
+
+    /// Calculate the hierarchical dendrogram for the givan linkage criterion.
+    pub fn dendrogram(&'a self, linkage: LinkageCriterion) -> Box<Dendrogram<&'a P>> {
+        let mut x = self.merge(linkage, INFINITY);
+        assert!(x.len() == 1);
+        let x = x.pop().unwrap();
+
+        self.convert(&x)
+    }
+
+    fn convert(&'a self, root: &Dendrogram<usize>) -> Box<Dendrogram<&'a P>> {
+        Box::new(match root {
+            &Dendrogram::Branch(d, ref a, ref b) =>
+                Dendrogram::Branch(d, self.convert(a), self.convert(b)),
+            &Dendrogram::Leaf(i) => Dendrogram::Leaf(self.points[i])
+        })
+    }
+
+    fn merge(&'a self, linkage: LinkageCriterion, threshold: f64) -> Vec<Box<Dendrogram<usize>>> {
+        let mut clusters: Vec<_> = (0..self.points.len()).map(|i|
+            Box::new(Dendrogram::Leaf(i))).collect();
+
+        // first abort criterion, there must be two clusters to merge them
+        while clusters.len() > 1 {
+            let (d, i, j) = (0..clusters.len())
+                .flat_map(|i| (i+1..clusters.len()).map(move |j| (i, j)))
+                .map(|(i, j)| (self.link(linkage, &clusters[i], &clusters[j]), i, j))
+                .fold((INFINITY, 0, 0), |a, b| if a.0 < b.0 { a } else { b });
+
+            // second abort criterium, max cluster distance
+            if d > threshold {
+                break;
+            }
+
+            let a = clusters.swap_remove(max(i, j));
+            let b = clusters.swap_remove(min(i, j));
+
+            clusters.push(Box::new(Dendrogram::Branch(d, a, b)));
+        }
+
+        clusters
+    }
+
+    #[allow(non_snake_case)]
```

Comment:
Not sure what you mean exactly, could you tell me which line you are referring to?